imhamt/src/bitmap.rs

```diff
@@ -0,0 +1,132 @@
+//! helpers and type for the HAMT bitmap
+//!
+//! The bitmap map a index into the sparse array index
+//!
+//! The number of bits set represent the number of elements
+//! currently present in the array
+//!
+//! e.g for the following elements and their indices:
+//!
+//! ```text
+//!     [ (0b0010_0000, x) ]
+//! ```
+//!
+//! will map into this bitmap:
+//!
+//! ```text
+//!     0b0010_0000 (1 bit set)
+//! ```
+//!
+//! and a vector of 1 element containing x
+//!
+//! ```text
+//!     | x |
+//! ```
+//!
+//! or the following elements and their indices:
+//!
+//! ```text
+//!     [ (0b0010_0000, x), (0b1000_0000, y), (0b0000_0010, z) ]
+//! ```
+//!
+//! will map into this bitmap:
+//!
+//! ```text
+//!     0b1010_0010 (3 bit set)
+//! ```
+//!
+//! and a vector of 3 elements containing x, y, z in the following order:
+//!
+//! ```text
+//!     | z | x | y |
+//! ```
+//!
+//!
+
+use super::hash::LevelIndex;
+
+/// This is a node size bitmap to allow to find element
+/// in the node's array
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct SmallBitmap(u32);
+
+impl SmallBitmap {
+    /// Create a new bitmap with no element
+    pub const fn new() -> Self {
+        SmallBitmap(0u32)
+    }
+
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.0 == 0
+    }
+
+    #[inline]
+    pub fn present(&self) -> usize {
+        self.0.count_ones() as usize
```

Comment:
BTW, we may want to pass `-C target-feature=+popcnt` to the compiler or add `#[target_feature(enable = "popcnt")]` to this function to get the compiler to emit code like
```
        popcntq %rdi, %rax
```
rather than
```
        movq    %rdi, %rax
        shrq    %rax
        movabsq $6148914691236517205, %rcx
        andq    %rax, %rcx
        subq    %rcx, %rdi
        movabsq $3689348814741910323, %rax
        movq    %rdi, %rcx
        andq    %rax, %rcx
        shrq    $2, %rdi
        andq    %rax, %rdi
        addq    %rcx, %rdi
        movq    %rdi, %rax
        shrq    $4, %rax
        leaq    (%rax,%rdi), %rax
        movabsq $1085102592571150095, %rcx
        andq    %rax, %rcx
        movabsq $72340172838076673, %rax
        imulq   %rcx, %rax
        shrq    $56, %rax
```

Reply:
ha, I had no idea that it was not automatically going to popcnt (I assumed llvm would just do that). I added the attribute to the function directly

Reply:
It's probably because it requires a CPU newer than ~11 years ago.